Add data validation

DougBurke opened the pull request:
# Summary

Improve the validation of arguments when creating Data objects.

Arrays sent to `Data` objects are now converted to ndarrays rather than sometimes being converted and sometimes not. The independent axis is now made read-only, so the only way to change a value is to replace the whole axis, in which case any existing filter or spatial region is removed. The size of a data object is now fixed, so once a field - such as the independent, dependent, or staterror columns - is set the size can not be changed. This means that a new object will have to be created rather than shortening or extending the fields of an existing object. Data objects have now gained a ndim attribute and this is used to prevent a number of error cases, such as trying to evaluate a 2D model on a 1D dataset.

A number of calls will now error out - normally raising a `sherpa.utils.err.DataErr` exception - when they are sent an argument of the wrong size whereas previously the call would have succeeded, but left the data object in an invalid state.

# Details

Fix #1470 and a number of other issues. [Aside: I had copied some of these commits into #1475 to try and reduce the burden of this PR, but I've just closed that as it has diverged enough from changes I've made here to make it hard to manage]. The code has been split up into a large number of small steps, as it makes it easier to see what such a small change makes to the test suite. I've spent a lot of effort breaking up the commit in this way but there are many ways to do this, and this is not guaranteed to be a minimal set of changes. There's also the fact that some changes will end up being re-worked in later commits (normally because of later changes), so please be aware of this when reviewing commits individually.

The aims are to move the validation to as "high-up" the chain as possible - so, for instance, the independent axis checks are made in the underlying `Axis` class (which was not-at-all-obvious to me when I started this) rather than in the `Data.__init__` method. For example, the validation does not generally appear in the ui code (there's always some place where you do need to do this).

There is an attempt to make the error messages include contextual information - that is, not "there is a mismatch between foo and bar" but "The sizes of foo and bar do not match: 23 vs 1024" - but there are cases when more (or less) context could be added and it is always going to be a trade off.

One change made in this PR is that the size of a `Data` object is, once set (that is, a field such as the independent axis, y, or staterror is set), can not be changed. This is a breaking change, but I argue that it is better for the code to be re-written to create a new data object rather than re-size an existing object. There is actually one test that I re-wrote because of this change.

As part of this work, `Data` objects now get `ndim` and `size` attributes, which make it easier to check these values. The`len` command can be used on `Data` obects: it returns the `size` value (unless it's not set, in which case 0 is returned). The `ndim` change works well to extend the support we added to models in #900. There's an example below showing this.

I've used the label "related fields" to represent the dependent and other fields (so `staterror`, `syserror`, and for PHA data the `grouping`, `quality`, `bin_lo`, `bin_hi`, `areascal`, `backscal`). We use this to enforce the size constraints and,for the PHA code, we can also do type conversions (`grouping` and `quality` should be `int16`) or support fields that can be a scalar or array (`areascal` and `backscal`). As part of this I've tried to add some documentation for the `bin_lo`/`bin_hi` fields, based on our discussion in #1471 

I have made the independent axis read-only, in the sense that the only way to change an independent axis value is to change the whole array. That is, while you can still do (where `data` is a `Data` object):

    data.dep[23] = 20
    data.staterror[mask] = 0.1

you can no longer say

    data.indep[0][23] = 45

(noting, as discussed below in the examples, that `indep` is actually a tuple, even in the 1D case, hence the need to access the first element of it here). To simplify this you can use the `x`/`x0`/... fields, that is

    data.x[23] = 45

for a `Data1D` dataset used to allow you to change the independent axis, but it now raises a `ValueError` letting you know the `assignment destination is read-only`. So, to change the independent axis you now need to do something like

    orig = data.x.copy()
    orig[23] += 3
    data.set_indep((orig, ))

[you can not change `.x` as that errors out, even with the `head` branch]. Because of this I have now made the code clear any existing mask or filter when the independent axis. See #1437 for one example of why we want to do this. It is something that we could change, but I like this approach, so be prepared! My argument is that if the independent axis is changed then we can not know how it was changed, so it's safest to clear any filters, and tell the user that they need to store/re-create the filter if they know how. Surprisingly enough the DataIMG case adds some surprising complexity here, since whenever the coord setting is changed the independent axis changes, but this is a "known" change so we **could** keep the filter, but in this case we do not know how to change the region filter string, so we can not keep the data consistent [*].

[*] technically we could support region filters that we can change between logical and physical systems (as it's just an affine transform), but this would require significant changes to the region interface to better match current CIAO capabalities, and it all goes out of the window once the world coordinate system is included.

The changes to the independent axis would seem no-different that the related fields, other than making them read only, but it lead to a number of surprising changes since

- the independent axis is actually stored in a separate object, and the `Data` objects delegate to it
- this separate object has no common ancestor, so in particular the 1D and 2D code paths are similar just by "accident" (i.e. the classes are intentionally similar); however, there were differences that caused differences in behavior until I realized this and had to adjust the 2D case (in particular to use the Axis class, as I'm about to discuss)
- the underlying axis object (Axis) has been split out to a meta class (Axis) and then concrete classes (PointAxis and IntegratedAxis) as it made more sense to me (but not really needed by the validation changes)
- for some reason the PSF code was using a 2D axis object (`EvaluationSpace2D`) even when dealing with 1D data, so some work needed to be done in the instrument code

The `size` attribute added to the `Data` objects in this PR defines the number of elements in the independent axis (or equivalently the number of elements in the dependent axis). However, the PHA case - with grouped data - introduces an "effective size" concept (the number of bins after the data has been grouped). Is there any way to take advantage of this? The trouble is that without adding this concept to the base `Data` object, we are always going to have to treat PHA data differently. We could also track the filtered size, as there are times when this is nice to know, but I think these are for a follow-up discussion.

Although this adds a lot of validation, we don't catch all cases: as an example of this is that of the `bin_lo` and `bin_hi` fields. We do not check that the units are sensible (e.g. that bin_hi[i] > bin_lo[i] and that bin_lo[i + 1] < bin_lo[i] which I think is meant to hold but I can not guarantee it), or that we do not check we have either neither set, or both set.

The validation is focused on the main `Data` objects, and I have not spent a lot of time on the ancillary data types - in particular the ARF and RMF. These extra types are an interesting case, as they do change some of the basic assumptions of the data class (e.g.that there's meaningful stat/syserror fields).

I have removed the warning about some values gettnig converted to `ndarray`. This is only an issue for the explicit `Dataxxx` calls since users using the `ui.load_arrays` call already have the values converted for you.

I have added documentation, both to the docstrings and the rtd pages.

# Tests [not updated to match the latest version but the change in number of tests should be similar]

With astropy, matplotlib, XSPEC 12.12.0, and ds9 I get

```
============================ 6939 passed, 32 skipped, 77 xfailed, 4 xpassed in 632.24s (0:10:32) =============================
```

This can be compared to the output of #1474 (after the tests have been added but before the code changes in this PR):

```
============================ 6917 passed, 32 skipped, 91 xfailed, 4 xpassed in 590.10s (0:09:50) =============================
```

Note that the `head` branch has 88 `xfailed` runs, so we have fixed some of these cases as well.

# Discussion points

This section is basically issue #1470 - I think our discussion of this PR has agreed with all of these, but I'm leaving it in (in particular, the "clear mask/region filter" part has probably not been thought about as much as the other parts?).

- does this make sense
- can we remove the warnings about converting inputs to ndarray now
- do we want to make the independent axis read-only (that is, remove the changes in this PR)
  - even if we do, do we want to clear the mask/region filter when the axis is changed (that is, remove the changes in this PR)

# Examples

Some examples use the ui API but you'd see something similar if you created the data object directly, or change the fields directly.

## creating a Data1D dataset

```python
import numpy as np
from sherpa.astro import ui
x = np.asarray([1, 2, 3])
y = np.asarray([1, 2, 3, 4])
ui.load_arrays(1, x, y)
```

This now errors out with the following, when you used to be able to create a mis-matched object:

```
sherpa.utils.err.DataErr: size mismatch between independent axis and y: 3 vs 4
```

## creating a dataset with 2D arrays

I spent a long time wondering why `ui.load_arrays(1, <2d array>, <2d array>, <2d array>)` did not error out, but that's because the `load_arrays` call flattens the arguments (losing some data) rather than error-ing out. This turns out to be done in the `io` code (for reasons), so I'm not going to investigate that code path.

```python
yy, xx = np.mgrid[1:4, 1:3]
zz = xx + yy

ui.Data2D("fall-over", xx, yy, zz)
```

This now fails with a `DataErr` reporting

```
Array must be 1D
```

This is one place where we could provide more context in the error message, but I have decided against it. The previous code errored out with

```
TypeError: Data arrays should be 1-dimensional. Did you call 'flatten()' on [[1 2]
 [1 2]
 [1 2]]?)
```

## changing a field to the wrong size

```python
x = np.asarray([10, 20, 30, 40])
y = np.asarray([15, 45, 13, 30])
ui.load_arrays(1, x, y)

ui.set_dep([3, 4])
ui.set_staterror([3, 4, 5])
```

With CIAO 4.14, these two commands succeed. They now fail with errors (`DataErr`):

```
size mismatch between independent axis and y: 4 vs 2
size mismatch between independent axis and staterror: 4 vs 3
```

## changing a field to a 2D value

```python
x = np.asarray([1, 2, 3, 4])
y = np.asarray([1, 2, 3, 4]).reshape([2, 2])

ui.load_arrays(1, x, x)
ui.set_dep(y)
```

With CIAO 4.14 this succeeded  but you have no idea what it means. This now fails with a `DataErr` reporting

```
Array must be 1D
```

This is one place where we could provide more context in the error message, but I have decided against it.

## Setting the independent axis after the dependent axis

This is a corner case:

```python
from sherpa.data import Data1D

data = Data1D('empty', None, None)
```

Here's two new features:

```python
print(len(data))
print(data.size is None)
```

which returns

    0
    True

If we set the size:

```python
data.dep = np.asarray([1, 2])
```

we now get

```python
print(len(data))
print(data.size)
```

which returns

    2
    2


```python
# You can use either
data.indep = (np.asarray([1, 2, 3]), )
data.set_indep((np.asarray([1, 2, 3]), ))
```

This two last lines now fail with a `DataErr` reporting

```
independent axis can not change size: 2 to 3
```

In CIAO 4.14 they do not error out.

### indep axis is annoying

You may have noticed that we used a single-element tuple to `indep`. If you instead try what seems to be the more-obvious call:

```python
data.indep = np.asarray([1, 2, 3])
```

then before this PR you get a strange-looking `TypeError`:

```
__init__() takes 3 positional arguments but 5 were given
```

and now the `TypeError` is more understandable:

```
independent axis must be sent a tuple, not ndarray
```

This could have raised a `DataErr` but I decided to leave it as `TypeError` as this is an "expert" error.

## changing the mask

Note that PHA data adds extra complexity:

```python
x = np.asarray([10, 20, 30, 40])
y = np.asarray([1, 2, 3, 4])
ui.load_arrays(1, x, x)

ui.ignore(hi=15)
```

Which gives

```
>>> print(ui.get_dep())
[10 20 30 40]
```

and

```
>>> print(ui.get_dep(filter=True)
[20 30 40]
```

<this has not changed in this PR>.

We can manually change the `mask` field, so what happens if we use the wrong size:

```
data = ui.get_data()
```

```
>>> print(data.mask)
[False  True  True  True]
```

Changing `mask` does not error out in `head` but now does,

```
data.mask = [False, True]
```

will now raise  a `DataErr` reporting:

```
size mismatch between independent axis and mask: 4 vs 2
```

## model checking

We can now improve the checks to ensure that data and model dimensionality make sense:

```python
ui.clean()
x = np.asarray([1, 2, 3])
ui.load_arrays(1, x, x)

ui.set_source(ui.const2d.mdl)
ui.calc_stat()
```

The main branch will error out in the `calc_stat` call with a strange error (the exact details depends on the mis-match in dimensions):

```
TypeError: function missing required argument 'x1lo' (pos 3)
```

The code will now error out in the `set_source` call with a better error:

```
DataErr: Data and model dimensionality do not match: 1D and 2D
```

## calc_foo_sum/sum2d

Issue #1444 discussed problems because the ui `calc_data_sum` and `calc_data_sum2d` routines did not check that they were being used with the correct data dimensionality. The result that calls could succeed or fail, depending on what was being done.  We now error out so that after

```
x1, x0 = np.mgrid[1:4, 1:3]
y = x0 + x1
shape = x.shape

ui.load_arrays(1, x0.flatten(), x1.flatten(), y.flatten(), shape, ui.DataIMG)
ui.set_source(1, ui.gauss2d.mdl)
```

then calls to `ui.calc_data_sum`, `ui.calc_model_sum`, and `ui.calc_source_sum` will raise a `DataErr` reporting

```
data set '' does not contain 1-D data
```

and similarly for a 1D case the `sum2d` versions will report

```
data set '' does not contain 2-D data
```

We could change the error report (as it is confusing, deciding whether you want the "call" to fail because it expects a different dimensionality of data, or for it to fail because the data set it is being used on has the wrong dimensions; I don't really care which we do).

## PHA related example

```python
from sherpa.utils.logging import SherpaVerbosity

ui.clean()
with SherpaVerbosity("ERROR"):
    ui.load_pha(1, "sherpa-test-data/sherpatest/3c273.pi")

ui.set_grouping([1, 2, 3])
```

This does not error out on `main` but raises a `DataErr` with this PR:

```
size mismatch between independent axis and grouping: 1024 vs 3
```

## Changing the independent axis

The old code was a bit strange, as you could do some things but not others. I think it's a bit cleaner now.

```
x = np.asarray([1, 2, 3])
y = np.asarray([5, 7, 6])
data = Data1D("example", x, y)
```

As a check, we can change an individual element of a dependent field (in this case the dependent axis):

```
>>> data.dep[1] += 10
>>> print(data.get_dep())
[ 5 17  6]
```

If we try to change the independent axis, we can either use the `indep` field or the `x<label>` (for `Data1D` it's just `x`).

For `head`/CIAO 4.14 we get

```
>>> print(data.get_indep())
(array([1, 2, 3]),)
>>> data.indep[0][1] = 15
>>> print(data.get_indep())
(array([ 1, 15,  3]),)
```

whereas with this PR we can not change the `indep` field: we get

```
ValueError: assignment destination is read-only
```

Checking using the `x` field, we get similar behavior with `head`:

```
>>> print(data.x)
[ 1 15  3]
>>> print(data.get_indep())
(array([ 1, 15,  3]),)
>>> data.x[2] += 30
>>> print(data.x)
[ 1 15 33]
>>> print(data.get_indep())
(array([ 1, 15, 33]),)
```

and with this PR we get

```
ValueError: assignment destination is read-only
```

Given the above, you might be surprised with the following error with `head`:

```
data.x += 30
```

which fails with 

```
AttributeError: can't set attribute
 ```

With this PR this is now

```
ValueError: output array is read-only
```

Note that the following behavior is unchanged in this PR (this is with the `head` branch but you get the same with this PR):

```
>>> from sherpa.data import Data1D
>>> d = Data1D('x', [1, 2, 3], [1, 2, 3])
/home/dburke/sherpa/sherpa-master/sherpa/data.py:88: UserWarning: Converting array [1, 2, 3] to numpy array.
  warnings.warn(f"Converting array {array} to numpy array.")
>>> d.x = [4, 5, 6]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dburke/sherpa/sherpa-master/sherpa/utils/__init__.py", line 171, in __setattr__
    object.__setattr__(self, name, val)
AttributeError: can't set attribute
```

DougBurke commented:
Rebased as https://github.com/sherpa/sherpa/pull/1474 (and a lot of other changes) have been merged

DougBurke commented:
Rebased to fix up a test conflict.

DougBurke commented:
About to rebase again to address python 3.10 vs older test behavior differences. I'll do this in several batches of commits.

anetasie reviewed (commented):
Started  my review. I'm looking at this from the user side
and I still need to run some interactive tests.

anetasie reviewed (commented):
I have reviewed part of this PR by reading the main description and building the branch.
I used the branch to perform a few tests with analysis of 1D spectra and 2D images with psf.
I did not see any issues.

hamogu commented on sherpa/models/regrid.py:
```diff
@@ -112,36 +115,163 @@ def end(self):
+
+        nx = len(self._x)
+        if nx > 0:
+            self._is_ascending = self._x[-1] > self._x[0]
```
How about `x = [3, 2, 5]`? 
How about `np.all(np.sort(x) == x)`? Thinking about it, there should probably a check in the `Axis` base class that ensures that it's monotonic (either ascending or decending). Something like 
```
d=np.diff(x)
if not np.all(d > 0) or np.all(d < 0):
    raise ...
```

DougBurke replied:
```
x = [3, 4, 4]
```

is another case. However, I didn't want to add extra checks as I'm still not completely sure about what we really support, and decided not to rock the boat too much. I should go and check a 1D data array that is unordered - which is something I think we have supported in the past, even if some of the plots may appear less-than-endearing to the user in this case.

DougBurke replied:
I've created #1577 to track this, as I do not know the intent of the original code, or once we added the "regrid" support.

DougBurke replied:
In other words, I don't want to enforce a monotonic axis before we've had time to work through all the implications?

anetasie replied:
The case with a repeated independent axis should be handled as a specific case and I don't think we have implemented any case which would properly fit the data like  this `x = [3, 4, 4]`.

DougBurke replied:
Well, you can fit such a dataset now - the following is a very-simplified case where we fit a constant, so the independent coordinate does not matter, but technically you could fit something like gauss2d:

```
>>> from sherpa.ui import *
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'
>>> load_arrays(1, [2, 5, 2], [2, 6, 4])
>>> set_source(const1d.mdl)
>>> fit()
Dataset               = 1
Method                = levmar
Statistic             = chi2gehrels
Initial fit statistic = 2.96289
Final fit statistic   = 0.823007 at function evaluation 4
Data points           = 3
Degrees of freedom    = 2
Probability [Q-value] = 0.662653
Reduced statistic     = 0.411504
Change in statistic   = 2.13988
   mdl.c0         3.59539      +/- 1.77422     
>>> load_arrays(1, [100, 200, 300], [2, 6, 4])
>>> set_source(const1d.mdl2)
>>> fit()
Dataset               = 1
Method                = levmar
Statistic             = chi2gehrels
Initial fit statistic = 2.96289
Final fit statistic   = 0.823007 at function evaluation 4
Data points           = 3
Degrees of freedom    = 2
Probability [Q-value] = 0.662653
Reduced statistic     = 0.411504
Change in statistic   = 2.13988
   mdl2.c0        3.59539      +/- 1.77422     
```

hamogu commented on sherpa/models/regrid.py:
```diff
@@ -112,36 +115,163 @@ def end(self):
+        then the data axis is said to be empty. The axis can be in
+        ascending or descending order.
+    hi : array_like or None
+        The ending point of the axis. It must match `lo` (either `None`
```
```suggestion
        The ending point of the axis. The number of elements must match `lo` (either `None`
```

hamogu commented on sherpa/models/regrid.py:
```diff
@@ -273,29 +415,41 @@ class EvaluationSpace2D():
+        #
+        # This means that this class does not check that x and y (if set) have
+        # the same length.
+        #
```
This code looks well for regular grids, but I have concerns on irregular grids. 

Imagine the following integrated grid:
```
x0= [1, 2, 3, 1, 2, 3]
x1 = [1, .5, 1, 2, 2, 2]
```
What does that even mean? It really needs a spec of the interpolation method when evaluating the model; it's even worse in n-D. I'm not sure we have thought through all potential problems here. Should there be a requirement that integrated data is on a regular grid?

DougBurke replied:
I do not understand the 2D regriud code - e.g. see #840. It is being used (or was used?) for the CSC so we have to be careful not to break things, but I really just have not had the time, energy, or implication to try and resolve any of this.

anetasie replied:
Does the `regrid` code deal with model on a higher resolution grid than the input data?

DougBurke replied:
@anetasie - I don't understand your question?

hamogu commented on sherpa/ui/utils.py:
```diff
@@ -224,6 +224,111 @@ def _remove_obj_from_main(name):
+    if not ignore:
+        data.mask = val
+    else:
+        assert False, "D"
```
?

DougBurke replied:
Hmm, that looks like a debugging statement. I shall have to investigate (and probably just remove the `else` clause).

DougBurke replied:
Going to remove it as part of the next rebase

DougBurke replied:
It's been removed, but leaving the conversation only for now.

hamogu commented on sherpa/ui/utils.py:
```diff
@@ -6256,14 +6325,28 @@ def set_model(self, id, model=None):
+                    'a model object or model expression string')
 
+        # Fit(data, model) does the dimensionality validation. Note
+        # that we assume that any convolution-style model (e.g.  a PSF
```
```suggestion
        # that we assume that any convolution-style model (e.g. a PSF
```

hamogu commented:
To the discussion points:

> does this make sense

yes

> can we remove the warnings about converting inputs to ndarray now

We absolutely should remove them.

> do we want to make the independent axis read-only (that is, remove the changes in this PR)
        even if we do, do we want to clear the mask/region filter when the axis is changed (that is, remove the changes in this PR)

I agree with the approach here. I'm not sure I would have coded it up like that from scratch, but I can imagine many situations where "changing the independent axis" is conceptually easier (even it not in code) then "sorry, can't change that. Make a copy". This includes all coordinate transformations , which come in many forms in astronomy - WCS for images, adjusting the WCS, red-shifting a spectrum etc. Thus, I think we should stick with what's here.

DougBurke commented on docs/developer/index.rst:
```diff
@@ -724,13 +736,133 @@ current way that models are checked:
+`~sherpa.astro.data.DataPHA.bin_hi` fields are not checked to ensure
+that both are set, or that they are in descending order, that the
+``bin_hi`` value is always larger than the correspondnig ``bin_lo``
+value, or that there are no overlapping bins.
```
I think basically I felt that working out what to do with the `bin_lo`/`bin_hi` fields was well in the "future work" category as I don't understand the ramifications. I know you have been recently thinking about it (eg #1564) so we can track this for the future.

DougBurke commented on sherpa/astro/tests/test_astro_data.py:
```diff
@@ -2690,9 +2699,12 @@ def test_invalid_independent_axis(data_args):
     data = data_class(*args)
     indep = data.indep
-    with pytest.raises(TypeError):
+    with pytest.raises(DataErr) as err:
```
I have been recently moving to this style, but not yet backdated to this PR...

DougBurke replied:
I've changed a bunch of tests (hopefully including this one) but I make no guarantee that all cases have been changed!

hamogu replied:
Both are valid, it's just that one is easier than the other. So, I don't think you have to go back and change them all, I just wanted to suggest it.

DougBurke commented:
Initial work here has been to apply @hamogu's typo fixes. I've still to address the more-substantial comments and add in some tests for code that is missing tests (a lot of these are in existing code that has seen small changes, so this is code that's never been tested, but it would be good to address these when possible - they are mostly error conditions). I'm likely to rebase this to get the latest changes so I'll be adding the tests before the changes as a check (at least, that's the plan).

DougBurke commented:
Rght. I'm seeing some truly bizarre behavior here (`pytest` has decided to use `python 3.9` even though `python 3.10` is the installed version), so I'm going to rebase to pick up recent changes.

DougBurke commented on sherpa/astro/data.py:
```diff
@@ -1656,6 +1705,149 @@ def counts(self):
+                val = numpy.asarray(val, dtype=numpy.int16)
+            except TypeError:
+                raise DataErr("notanintarray") from None
+
```
We could use an unsigned integer type, but I'm a bit wary of doing that (eg the FITS standard has it as signed I think). For now I suggest not adding a restriction on the value.

hamogu replied:
Note that we can check >=0 without making it a signed integer, since unsigned integers could also be >0. But fine to not test here, too.

DougBurke commented on sherpa/instrument.py:
```diff
@@ -540,137 +725,147 @@ def __call__(self, model, session=None):
+            raise PSFErr('nopsf', self._name)
 
-        pixel_size_comparison = self._check_pixel_size(data)
+        # TODO: Should we treat origin as we do center and size?
```
I am **very** reluctant to add changes here as I **really** don't understand this code (and didn't write it). I think a clean up (or at least review)would make sense, but not in this PR.

hamogu replied:
Agreed.

DougBurke commented:
Currently in the process of addressing some comments, adding tests for untested code paths (mainly error handling changes), and then moving these changes around so the tests are written first (and some problematic code is never written). OI'll holler when this has been finished.

DougBurke commented on sherpa/astro/data.py:
```diff
@@ -1590,12 +1612,18 @@ def __init__(self, name, channel, counts, staterror=None, syserror=None,
+        #
         channel = _check(channel)
+        if channel is not None:
+            self._size = len(channel)
```
For all-but-PHA we have the number of elements as the size. For PHA data, do we want the size to be the number of channels or the number of grouped bins?

hamogu replied:
I suggest to use the number of channels as size for consistency.

DougBurke replied:
This has now been added to the `Data` class: it uses the size attribute (if set, otherwise returns 0), so ifor PHA data it does use the number of channels (not groups).

hamogu commented:
@DougBurke I see you are actively working on this and I know we'd like to get it into the next release, so please ping me when you think it's a good time to re-review.

DougBurke commented on sherpa/data.py:
```diff
@@ -488,10 +506,12 @@ def apply(self, array):
 
+        # Ensure we always return a ndarray.
+        #
+        array = numpy.asarray(array)
```
In a later check we wend up passing `array` to `_check`, which calls `npumpy.asanyarray`. In general we have a "passing regard" for numpy subclasses, as there's a number of places we just call `asarray` when we perhaps could use `asanyarray`. It would need some tests set up so we could do this properly, which is definitely a problem for another year ;-)

DougBurke commented on docs/developer/index.rst:
```diff
@@ -698,6 +698,18 @@ Notes
+``y`` for the independent and dependent axes, and a 2D dataset will
+have ``x0``, ``x1``, and ``y`` values, with each value stored as a 1D
+`ndarray`. This makes it easy to deal with filters and sparse or
+irregularly-placed grids.
```
Here's my current attempt

![screenshot](https://user-images.githubusercontent.com/224638/189700150-c2e02ee5-9517-45c0-b329-5e34a41608c7.png)

DougBurke commented on docs/developer/index.rst:
```diff
@@ -724,13 +736,133 @@ current way that models are checked:
+
+There is currently no easy way to identify whether a data object
+requires integrated (low and high edges) or point axes (the coordinate
+at which to evaluate the model).
```
Created #1575 to track this

DougBurke commented on sherpa/astro/tests/test_astro_data.py:
```diff
@@ -2953,8 +2974,12 @@ def test_pha_related_field_can_not_be_a_set_sequence(field):
+        setattr(data, field, [{"abc", False, 23.4}] * 3)
+
+    # once non-quality/grouping fields fail the error check will need to be changed
+    assert str(err.value) == "Array must be a sequence of integers or None"
```
I don't think we want syserror to be forced to be an integer, which is what this error message suggests. Aha. The syserror (in fact, most fields being tested here) actually do not trigger this warning - i.e they are marked as xfail. Aha, this gets resolved in a later commit so the error message is then "Array must be a sequence or None".

We could support setting the systematic error to a scalar (e.g. 0.05 to mean 5%, which then gets applied to the existing data) - and in fact I think the UI does support this (would need to check, but I seem to remember it doing so). There are issues to resolve: i.e. if I've set it to 0.05 and then I've changed the y/dependent axis, should it know to automatically apply this to the new data or require the call to be re-run (arguments are possible for all sides).

hamogu replied:
The latter (set it to the scalar 0.05) is what I had in mind. I need to read up on the docs, but intuitively, I had expected it to be an absolute number, not a fraction (i.e. 0.05 is 0.05, which can be a lot as in 0.01 +/- 0.05, or a little, as in 1234.566 +\- 0.05)
I don't think it's required to reset is when the dependent axis changes. To me, that's a property of the dataset. Since the independent axis is not harder to change, I would say that "I'm have a totally different measurement and I'm just re-using the same object" is not the prime use case we should be worried about.

hamogu replied:
That said, if it's much easier to reset it  when the dependent axis is changed, I can work with that, too. As @DougBurke and agument can be made either way, we just have to write down in the docs what choice we made.

DougBurke replied:
So, we have different behavior in the UI layer to the data object. In the UI layer you can call `set_staterror` or `set_syserror` to a scalar and it replicates the value for each bin [*]. If you manually change the `staterror` or `syserror` field of the data object then in CIAO 4.14 it just sets the value to a scalar, which technically may work (what with broadcasting and all) but this PR is all about trying to make the code a bit easier to write.

We could move the logic that's currently in `ui.set_staterror`/`ui.set_staterror` into the `staterror`/`syserror` properties (it would fit into my desire to reduce the logic in the ui code).

Here's an example from a CIAO 4.14 sherpa session:

```
sherpa-4.14.0> load_arrays(1, [1, 2, 3], [0, 1, -1])

sherpa-4.14.0> get_staterror()
array([1.8660254 , 2.32287566, 1.8660254 ])

sherpa-4.14.0> get_syserror()
DataErr: data set '1' does not specify systematic errors

sherpa-4.14.0> set_syserror(0.5)

sherpa-4.14.0> get_syserror()
array([0.5, 0.5, 0.5])

sherpa-4.14.0> set_staterror(1.2)

sherpa-4.14.0> get_staterror()
array([1.2, 1.2, 1.2])

sherpa-4.14.0> data = get_data()

sherpa-4.14.0> data.staterror = 2

sherpa-4.14.0> data.syserror = 3

sherpa-4.14.0> print(data)
name      = 
x         = Int64[3]
y         = Int64[3]
staterror = 2
syserror  = 3
```

[*] note to self: what does it do for grouped PA data

DougBurke replied:
If you look in `sherpa.ui/utils.py` (not going to link to it as it'll just get rebased away soon...) one of the things I did was move some code related to setting dependent axis/errors/filters into separate routines that could be called from the `Session` class. You can see below that we have logic in the ui layer to allow some of these fields to be set to a scalar, but they are not in the `Data` methods. I'd prefer to resolve this in a separate, targeted, PR rather than this one.

```
def set_dep(data, val):
    """Set the dependent axis.

    Parameters
    ----------
    data : sherpa.data.Data instance
    val : scalar or array_like
        If a scalar then use the same value for all elements.

    """

    if numpy.iterable(val):
        dep = numpy.asarray(val, SherpaFloat)
    else:
        val = SherpaFloat(val)
        dep = numpy.array([val] * data.size)

    data.dep = dep


def set_error(data, field, val, fractional=False):
    """Set the error field.

    Parameters
    ----------
    data : sherpa.data.Data instance
    field : {"staterror", "syserror"}
    val : None, a scalar, or array_like
        The statistical error.
    fractional : bool, optional
        If `val` is a scalar then set errors to
        `val * data.y`.

    Notes
    -----
    The `val` field, when an array, must match the data.size
    field, even for grouped PHA data.

    """

    if val is None:
        err = None

    elif numpy.iterable(val):
        err = numpy.asarray(val, SherpaFloat)

    else:
        val = SherpaFloat(val)
        if sherpa.utils.bool_cast(fractional):
            err = val * data.get_dep()
        else:
            err = numpy.array([val] * data.size)

    setattr(data, field, err)


def set_filter(data, val, ignore=False):
    """Set the filter field.

    Parameters
    ----------
    data : sherpa.data.Data instance
    val : array_like
        The filter array (bool-like).
    ignore : bool, optional
        If set the filter should be inverted.

    Notes
    -----
    The `val` field should match the mask array, that is the
    "effective" size of the data. This is only different from
    data.size when the data is grouped (and at least one group
    contains multiple channels).

    """

    val = numpy.asarray(val, dtype=numpy.bool_)
    nval = len(val)

    # Note that we do not use data.size as the check, as that fails
    # for grouped PHA data. Is there some way to have an "effective
    # size" value?
    #
    nexp = len(data.get_y(False))
    if numpy.iterable(data.mask):
        if nexp != nval:
            raise sherpa.utils.err.DataErr('mismatchn', 'data', 'filter',
                                           nexp, nval)

        if not ignore:
            data.mask |= val
        else:
            data.mask &= ~val

        return

    # The mask attribute checks the length, so we do not have to here.
    #
    if not ignore:
        data.mask = val
    else:
        data.mask = ~val
```

DougBurke commented on sherpa/data.py:
```diff
@@ -35,6 +35,54 @@
+
+- there is very-limited support for NumPy masked arrays, but it is
+  expected that any filtering has been applied before the `Data`
+  object is created.
```
Unfortuantely I don't understand what level of support we have for masked arrays from #793, so I don't know what text would be helpful here.

hamogu replied:
I *should* have an understanding of #793 and will suggest text. How about:

Numpy masked arrays can be used to initialize the dependent variable and the mask is converted to the format of the mask attribute of the `Data` object, taking into account that in :py:attr:`~sherpa.data.Data.mask` a value of `True` indicates a masked quantity, while the opposite is true in numpy.

hamogu replied:
Note my edit of the comment above 30 s after I posted it, because I confused again what `mask=True` means in sherpa and numpy. Of course...

DougBurke replied:
I've just added a commit to address this (more should probably be done, but something is better than nothing I feel).

DougBurke commented on sherpa/data.py:
```diff
@@ -295,44 +371,55 @@ def get(self, filter=False):
+        return self.x_axis.x, self.y_axis.x
+
+    # Should these be deprecated now that the code attempts to mimic EvaluationSpace2D?
+    #
```
In other words, I've left this comment for later review, but I don't feel it's a good thing to change here.

hamogu replied:
Make sense.

DougBurke commented on sherpa/data.py:
```diff
@@ -295,44 +371,55 @@ def get(self, filter=False):
+            A tuple representing the x0 and x1 axes.
         """
-        return self.x0, self.x1
+        return self.x_axis.x, self.y_axis.x
```
So I think there's confusion in our code over whether 2D data is always gridded or not. As you point out, the docstring for the `grid` method here assumes there is a grid, but it's not required (and I'd argue we wouldn't want to require it). 

I've created #1576 to track this, as I just don't know.

DougBurke commented on sherpa/instrument.py:
```diff
@@ -413,102 +468,231 @@ def _create_tail_grid(axis_list):
 
     def __init__(self, name='psfmodel', kernel=None):
+
+        # store the name without the leading "psfmodel." term that Model adds.
         self._name = name
+
         self._size = None
         self._origin = None
         self._center = None
         self._must_rebin = False
+        self._model = None
+
+        self._kernel = None
```
I am reluctant to make what-appear-to-be-simple clean-up changes in the instrument code as "it is complicated" and I don't feel we have a good-enough set of tests that cover the desired functionality (I've been adding tests but these are to check certain code paths, rather than being a meaningful set of tests that cover how users are expected to use these).

I note that a number of these fields are internal to the class, as they are used to populate properties (e.g. "size", "origin", "center") so the desire for a docstring is, I think, somewhat lessened from if these are user-visible fields.

I would expect any subclass to call this routine, but I can't guarantee it.

DougBurke commented:
@hamogu - I think I've finished addressing your comments. I've done a last pass through the PR message above (mainly subtle changes).

One thing I don't think we've touched on much is the sub-note I made

- do we want to make the independent axis read-only (that is, remove the changes in this PR) 
  - even if we do, do we want to clear the mask/region filter when the axis is changed (that is, remove the changes in this PR)

When you change the axis, should we reset the filter. I think it's more an issue for the `DataIMG` case, since whenever you change the `coord` setting you implicitly change the axes (for reasons), which is unlike the analysis/PHA case, but I think it makes sense (I would expect most people not to be swapping coord settings). It also does impact 1D cases, but there you have to explicitly change the axes. I note that there's no "optimisation" which checks to see whether the axis has actually changed - ie the equivalent of `set_indep(get_indep())` is considered to be "the axis has been changed".

hamogu commented:
I did say in https://github.com/sherpa/sherpa/pull/1477#issuecomment-1228908661 that I think we should stick with this or at least do something very similar for writing the independent axis. Let me expand on this: Conceptually, it is often useful to change the number on the independent axis when you do some coordinate transformation, e.g. keV-> Ang, wavelength -> red-shift, L_x-> Mass or pixels -> WCS. 

All those example suggest **not** to automatically clear the filter: If I mask out some wavelength bins because they have cosmics or ISM absorption lines in them, those lines are still going to be in the same bins after transforming from wavelength -> redshift. If I have an image, and I mask out a certain region because of a foreground object, then that same region should still be masked if I label the axis in RA/DEC instead of x/y.

With this "change the independent axis is a change of reference system", one would want to keep the same mask. So filters should not be kept at the same numerical value (it might make sense to mask out x=234:345 , but not RA=234:345), but what we should keep is the mask, i.e. if bin 45-57 are masked out, then be keep masking out bin 45-57, even though those bins now correspond to a different numerical value (and unit) on the independent axis.

This is a choice of design and it's always possible to work around it (`old_mask = data.mask.copy();data.x=np.arrange(123); data.mask=old_mask`), but I do have a preference to keep the mask as is when the independent axis is changed. In my mind, that's a "same data, different unit" situation. If I want an entirely new independent axis wit ha new mask, I should make a new data object.

hamogu commented:
That would also mean that `set_indep(get_indep())` would be a no-op.

Should be finish discussing this before I re-review the entire PR? It's now > 15000 lines of code changed and it's so different I probably need to look at it all again. I'll do it - but it's not done quickly.

DougBurke commented:
Background info, not sure it is quite relevant to what you are talking about @hamogu but I'm busy with kind-of-actual-science today.

So, for the 2D filter case we don't have a way to convert the mask back into a filter expression as we do with the 1D case. That means we carry around the actual filters the user has requested (which, in some sense, is nicer than the 1D case), which means that when you change the coordinate system the "stored" mask is now in the wrong coordinate system [*]

[*] CIAO's region module has more functionality than the one we include in Sherpa, and it *could* convert between simple transform changes (change of scale or offset) but not really the generic WCS changes we technically allow.

hamogu commented:
So, for the 2D case, we have no choice but clear the filter, because we just don't have the ability to keep it.
Does that means we want to clear it in the 1D case also for consistency? Do we even agree that it might be useful to keep the filter?

DougBurke commented:
Here's the text I wrote in the commit message

> Because of this I have now made the code clear any existing mask or filter when the independent axis. See #1437 for one example of why we want to do this. It is something that we could change, but I like this approach, so be prepared! My argument is that if the independent axis is changed then we can not know how it was changed, so it's safest to clear any filters, and tell the user that they need to store/re-create the filter if they know how. Surprisingly enough the DataIMG case adds some surprising complexity here, since whenever the coord setting is changed the independent axis changes, but this is a "known" change so we could keep the filter, but in this case we do not know how to change the region filter string, so we can not keep the data consistent [*].
>
> [*] technically we could support region filters that we can change between logical and physical systems (as it's just an affine transform), but this would require significant changes to the region interface to better match current CIAO capabalities, and it all goes out of the window once the world coordinate system is included.

So it's a call on our side to make: it's a balance between code complexity [#] and user expectation [$].

[#] actually, in this case I don't think it makes much difference, and not clearing the filter is almost-certainly less complexity

[$] I don't think there's a compelling argument for other side here

Although clearing the filter for images (which we can do independently of the 1D case) would help some users, it is also not something we absolutely must do (as #1437 tries to make clear, we have options).

I shall look to see what removing the code would entail, in case we decide to do that.

hamogu commented:
@anetasie What do you think? See 
https://github.com/sherpa/sherpa/pull/1477#issuecomment-1245382892 and the following few comments. Should the filter always be cleared when the independent axis is changed?

@DougBurke Even if we decide to go that rout, it might be the easiest to merge this PR as is and then do a follow-up where we add the "keep the filter in 1D" since that's a pretty easy and isolated change. Please don't waste your time trying to back it out here - and certainly don't do that before we even decided that's what we want.

hamogu commented on docs/developer/index.rst:
```diff
@@ -698,6 +698,51 @@ Notes
+
+  >>> from sherpa.data import Data1D, Data1DInt, Data2D
+
+As examples, we have a one-dimension dataset with data values
```
```suggestion
As examples, we have a one-dimensional dataset with data values
```

hamogu commented on docs/developer/index.rst:
```diff
@@ -698,6 +698,51 @@ Notes
+
+An "integrated" one-dimensional dataset for the independent axis
+bins 23-44, 45-50, 50-53, and 55-57, with data values of
+12, 14, 2, and 22:
```
```suggestion
12, 14, 2, and 22 looks like this:
```
Just making that a full sentence with a verb.

hamogu commented on docs/developer/index.rst:
```diff
@@ -698,6 +698,51 @@ Notes
+
+  >>> x1, x0 = np.mgrid[20:30:2, 5:20:2]
+  >>> shp = x0.shape
+  >>> x0 = x0.flatten()
+  >>> x1 = x1.flatten()
+  >>> y = np.sqrt((x0 - 10)**2 + (x1 - 31)**2)
+  >>> d4 = Data2D("ex4", x0, x1, y, shape=shp)
```
If we have 2D data for x, we typically also have 2D data for y.  Here, we just make that data for the example, but in a typical use case, we would get the x  and y in some other way (e.g. read from a file) and they wuld both be 2d to begin with.
```suggestion
  >>> y = np.sqrt((x0 - 10)**2 + (x1 - 31)**2)
  >>> x0 = x0.flatten()
  >>> x1 = x1.flatten()
  >>> y = y.flatten()
  >>> d4 = Data2D("ex4", x0, x1, y, shape=shp)
```

DougBurke replied:
I don't really see this as an improvement but as I wrote it I may biased...

hamogu commented on docs/developer/index.rst:
```diff
@@ -724,13 +769,135 @@ current way that models are checked:
   ``ndim`` value (ignoring those models whose dimensionality
   is set to ``None``). If there is a mis-match then a
   :py:class:`sherpa.utils.err.ModelErr` is raised.
+* as :ref:`descibed below <data_design_ndim>`, the dimensions of
```
```suggestion
* as :ref:`described below <data_design_ndim>`, the dimensions of
```

hamogu commented on docs/developer/index.rst:
```diff
@@ -724,13 +769,135 @@ current way that models are checked:
+
+Prior to Sherpa 4.14.1, the `~sherpa.data.Data` object did not have
+many explicit checks on the data it was sent, instead relying on
+checks when the data was used. This has been changed to apply
+validation checks when fields are changed, rather than when the data
```
Just avoiding to start several sentences in a raw with "This has been"
```suggestion
checks when the data was used. Now, validation checks  are done
when fields are changed, rather than when the data
```

hamogu commented on docs/developer/index.rst:
```diff
@@ -724,13 +769,135 @@ current way that models are checked:
+`~sherpa.astro.data.DataPHA.quality` columns for PHA data which
+are converted to integer values.
+
+There is limited support for NumPy masked arrays, but the assumption
+is that and data has been masked before being added to a
+`~sherpa.data.Data` object.
+
```
The sentences in borked (technical term in British English) and it's also explained elsewhere, so just delete?

```suggestion

```

hamogu commented on docs/developer/index.rst:
```diff
@@ -724,13 +769,135 @@ current way that models are checked:
+is that and data has been masked before being added to a
+`~sherpa.data.Data` object.
+
+One example of where there is incomplete validation is that the
```
```suggestion
One example incomplete validation is that the
```

DougBurke replied:
Did you mean

```suggestion
One example of incomplete validation is that the
```

hamogu replied:
Yours sounds better ;-)

wmclaugh commented on docs/developer/index.rst:
```diff
@@ -811,18 +980,25 @@ In this case the term ``32424.43`` is converted to an
+creating a model that can "wrap" another model. The wrapping model
+will evaluate the wrapped model on the requested grid, and then apply
+any modifications.  Examples include the
+:py:class:`sherpa.instrument.PSFModel` class, which creats
```
```suggestion
:py:class:`sherpa.instrument.PSFModel` class, which creates
```

wmclaugh commented on docs/developer/index.rst:
```diff
@@ -811,18 +980,25 @@ In this case the term ``32424.43`` is converted to an
+:py:class:`sherpa.astro.xspec.XSConvolutionKernel` class, which
+creates :py:class:`sherpa.astro.xspec.XSConvolutionModel` instances.
+
+Wen comnining models, :py:class:`~sherpa.models.model.BinaryOpModel`
```
```suggestion
When combining models, :py:class:`~sherpa.models.model.BinaryOpModel`
```

hamogu replied:
```suggestion
When combining models, :py:class:`~sherpa.models.model.BinaryOpModel`
```

hamogu commented on docs/developer/index.rst:
```diff
@@ -811,18 +980,25 @@ In this case the term ``32424.43`` is converted to an
-class, which creates :py:class:`sherpa.astro.xspec.XSConvolutionModel`
-instances.
+For those models that require the full set of elements, such as
+multiplication by a :term:`RMF` or a convolution kernel, requires
```
sentence borked. Not sure my suggestion fixes it though.
```suggestion
multiplication by a :term:`RMF` or a convolution kernel, this requires
```

I realize that this was text that's just reformatted and not written here.

hamogu commented on docs/developer/index.rst:
```diff
@@ -811,18 +980,25 @@ In this case the term ``32424.43`` is converted to an
+
+Wen comnining models, :py:class:`~sherpa.models.model.BinaryOpModel`
+(actually, this check is handled by the super class
+:py:class:`~sherpa.models.model.CompositeModel`), will ensure that the
```
```suggestion
:py:class:`~sherpa.models.model.CompositeModel`), this approach will ensure that the
```

hamogu commented on sherpa/astro/data.py:
```diff
@@ -45,6 +45,23 @@
+------
+
+The `DataPHA` class adds support for grouping data - effectively
+reducing the number of values in a data set - and adding an extra way
```
```suggestion
reducing the number of values in a data set - and adds an extra way
```

hamogu commented on sherpa/astro/data.py:
```diff
@@ -1807,6 +1856,149 @@ def counts(self):
+        # rather than iterable, to avoid selecting strings).
+        #
+        if self.grouped and val is not None and not numpy.isscalar(val):
+            # The assumption is that is the data is grouped then it contains data.
```
And now for the important suggestions...
```suggestion
            # The assumption is that if the data is grouped then it contains data.
```

hamogu commented on sherpa/astro/data.py:
```diff
@@ -1987,21 +2181,22 @@ def get_specresp(self, filter=False):
-            specresp = arf.get_dep()
-            elo, ehi = arf.get_indep()
-            lo, hi = self._get_ebins(group=False)
+        # Should we allow an ARF-only analysis?
```
Github does not let me make a suggestion in lines not edited here, but with this discussion in mind, I suggest the following addition to the docstring:

This function only works for datasets where ARF and RMF are separate. In same cases (e.g. XMM/RGS), the effective area and the response are combined into a single matrix and this function cannot extract the effective area from that.

hamogu reviewed (changes requested):
I'm having a closer look at the docs now. Of course, that's all small potatoes compared to the code changes.

hamogu commented:
Otherwise, I think the code changes are in good shape.

DougBurke commented:
The remaining issue here that we have been talking about is related to the following difference in behavior between CIAO 4.14 and this branch. The 4.14 version is

```
sherpa-4.14.0> load_arrays(1, [1, 2, 4], [3, 4, 5])

sherpa-4.14.0> get_filter()
'1.0000:4.0000'

sherpa-4.14.0> get_data().set_indep(([1, 2, 3], ))

sherpa-4.14.0> print(get_data().x)
[1 2 3]

sherpa-4.14.0> get_filter()
'1.0000:3.0000'

sherpa-4.14.0> load_arrays(1, [1, 2, 4], [3, 4, 5])

sherpa-4.14.0> get_filter()
'1.0000:4.0000'

sherpa-4.14.0> notice(lo=3)

sherpa-4.14.0> get_filter()
'4.0000'

sherpa-4.14.0> get_data().mask
array([False, False,  True])

sherpa-4.14.0> get_data().set_indep(([1, 2, 3], ))

sherpa-4.14.0> get_filter()
'3.0000'

sherpa-4.14.0> get_data().mask
array([False, False,  True])
```

So, I have changed the independent axis from [1, 2, 4] to [1, 2, 3] and the filter mask has not been changed.

With this PR we get (I've just restricted it to the changes that "matter"):

```
>>> load_arrays(1, [1, 2, 4], [3, 4, 5])
>>> get_filter()
'1.0000:4.0000'
>>> notice(lo=3)
>>> get_filter()
'4.0000'
>>> get_data().set_indep(([1, 2, 3], ))
/home/dburke/sherpa/sherpa-work/sherpa/data.py:1060: UserWarning: Filter has been removed from ''
  warnings.warn(f"Filter has been removed from '{self.name}'")
>>> get_filter()
'1.0000:3.0000'
>>> get_data().mask
True
```

When changing the independent axis we have now cleared the filter/mask.

This looks strange here for the 1D case. It is a bit-more hidden and hence surprising for the 2D/image case. Here's the 4.14 version:

```
sherpa-4.14.0> load_image("acisf07999_000N001_r0035_regevt3_srcimg.fits")

sherpa-4.14.0> get_coord()
'logical'

sherpa-4.14.0> ignore2d("circle(100,100,10)")

sherpa-4.14.0> get_filter()
'Field()&!Circle(100,100,10)'

sherpa-4.14.0> set_coord("physical")

sherpa-4.14.0> get_filter()
'Field()&!Circle(100,100,10)'
```

We can see that the filter has not changed, but it's no-longer relevant.

With this PR we get

```
>>> load_image("acisf07999_000N001_r0035_regevt3_srcimg.fits")
>>> get_filter()
''
>>> get_coord()
'logical'
>>> ignore2d("circle(100,100,10)")
>>> get_filter()
'Field()&!Circle(100,100,10)'
>>> set_coord("physical")
/home/dburke/sherpa/sherpa-work/sherpa/astro/data.py:4819: UserWarning: Region filter has been removed from 'acisf07999_000N001_r0035_regevt3_srcimg.fits'
  warnings.warn(f"Region filter has been removed from '{self.name}'")
>>> get_filter()
''
```

For various reasons, the `set_coord` call changes the independent axis of the image class, so we now warn/clear the filter.

Do we do this or not?

hamogu commented:
I like the old behavior for 1 D datasets (as I've argued above), but not the non-sense that happens in 2D datsets. The new warning says what happens. 
So, here is a suggestion: For 1 D datasets, we revert to the 4.14 behavior, for 2D datasets we remove the filter with the warning as in this PR. That's not fully consistent, but I don't think it's a big problem. We can declare that we in principle like the old behavior, it's just not implemented in 2D and thus we have a warning. That allows us to add it later (and remove the warning) if we want to. 
For 1D everything will stay the same as in 4.14, so there is no risk of breaking any users workflow. For 2D we are adding a warning and removing a filter that made no sense in 4.14, so that's a change that makes things better, i.e. a bug fix. Call is a pragmatic suggestion. I don't think it's that hard to do in code and keep backwards=compatible where it makes sense. It's not any less confusing than it was before and I don't think many people use 1D and 2D datasets interchangably. The filtering works differently anyway (since the regions library is used for 2D), while in 1D you can pass in a mask, and thus I think that's OK.

I note however that I'm not going to insist on it - I rarely change the independent axis, so I won't encounter this much in practice.

@anetasie ? @DougBurke ? Sounds like we just have to agree on something.

hamogu reviewed (approved):
At last, I think this is all done. Let's get it in!

hamogu commented:
Congratulations! This PR also brings us a significant reduction of the number of open issues - that's something to be proud of @DougBurke !
